modules/vault/src/test/java/org/apache/ignite/internal/vault/impl/VaultBaseContractsTest.java

```diff
@@ -0,0 +1,225 @@
+
+        VaultEntry v = storage.get(key).get();
+
+        assertFalse(v.empty());
+        assertEquals(val, v.value());
+
+        storage.put(key, val);
+
+        v = storage.get(key).get();
+
+        assertFalse(v.empty());
+        assertEquals(val, v.value());
+    }
+
+    /**
+     * remove contract.
+     */
+    @Test
+    public void remove() throws ExecutionException, InterruptedException {
+        ByteArray key = getKey(1);
+        byte[] val = getValue(key, 1);
+
+        assertNull(storage.get(key).get().value());
+
+        // Remove non-existent value.
+        storage.remove(key);
+
+        assertNull(storage.get(key).get().value());
+
+        storage.put(key, val);
+
+        VaultEntry v = storage.get(key).get();
+
+        assertFalse(v.empty());
+        assertEquals(val, v.value());
+
+        // Remove existent value.
+        storage.remove(key);
+
+        v = storage.get(key).get();
+
+        assertNull(v.value());
+    }
+
+    /**
+     * range contract.
+     */
+    @Test
+    public void range() throws ExecutionException, InterruptedException {
+        ByteArray key;
+
+        Map<ByteArray, byte[]> values = new HashMap<>();
+
+        for (int i = 0; i < 10; i++) {
+            key = getKey(i);
+
+            values.put(key, getValue(key, i));
+
+            assertNull(storage.get(key).get().value());
+        }
+
+        values.forEach((k, v) -> storage.put(k, v));
+
+        for (Map.Entry<ByteArray, byte[]> entry : values.entrySet())
+            assertEquals(entry.getValue(), storage.get(entry.getKey()).get().value());
+
+        Iterator<VaultEntry> it = storage.range(getKey(3), getKey(7));
+
+        List<VaultEntry> rangeRes = new ArrayList<>();
+
+        it.forEachRemaining(rangeRes::add);
+
+        assertEquals(4, rangeRes.size());
+
+        //Check that we have exact range from "key3" to "key6"
+        for (int i = 3; i < 7; i++)
+            assertEquals(values.get(getKey(i)), rangeRes.get(i - 3).value());
+    }
+
+    /**
+     * watch contract.
+     */
+    @Test
+    public void watch() throws ExecutionException, InterruptedException {
+        ByteArray key;
+
+        Map<ByteArray, byte[]> values = new HashMap<>();
+
+        for (int i = 0; i < 10; i++) {
+            key = getKey(i);
+
+            values.put(key, getValue(key, i));
+        }
+
+        values.forEach((k, v) -> storage.put(k, v));
+
+        for (Map.Entry<ByteArray, byte[]> entry : values.entrySet())
+            assertEquals(entry.getValue(), storage.get(entry.getKey()).get().value());
+
+        AtomicInteger counter = new AtomicInteger();
+
+        VaultWatch vaultWatch = new VaultWatch(changedValue -> counter.incrementAndGet());
+
+        vaultWatch.startKey(getKey(3));
+        vaultWatch.endKey(getKey(7));
+
+        storage.watch(vaultWatch);
+
+        for (int i = 3; i < 7; i++)
+            storage.put(getKey(i), ("new" + i).getBytes());
+
+        Thread.sleep(500);
```

Comment:
Maybe replace counter and sleep with CountDownLatch?

Reply:
I agree, every test should (hopefully) fit into 50-100 milliseconds duration, 5-10 ideally, 500 it too long.

Reply:
+1 
I think we should avoid `Thread.sleep()` in tests.